### PR TITLE
refactor lookup plugins using top down method

### PR DESCRIPTION
**Pull request by mikemorency**
##### SUMMARY
This change rewrites the lookup plugin logic to use a top down approach to finding objects. The plugins will now step through an objects path and try to lookup each intermediate object until it gets to the final object.

The old method has a few issues:
1. The existing plugins use an incorrect search method  which yield inconsistent results (as described in https://github.com/ansible-collections/vmware.vmware_rest/issues/500)
2. They also cannot find resources with the same name, even if the paths given are unique. For example, if `/datacenter/vms/my-folder` and `/datacenter/hosts/my-folder` both exist, you can never find either one
3. Items with the same name will cause conflicts, even if those items have slightly different paths. For example: `/datacenter/vms/foo` and `/datacenter/hosts/foo`  existing at the same time makes it difficult to search for either folder

Fixes:
https://github.com/ansible-collections/vmware.vmware_rest/issues/500
https://github.com/ansible-collections/vmware.vmware_rest/issues/445
https://github.com/ansible-collections/vmware.vmware_rest/issues/324

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
all lookup plugins

##### ADDITIONAL INFORMATION
The original top down approach is described in https://github.com/ansible-collections/vmware.vmware_rest/issues/445. I tried to get that version working but ran into a few issues.

### CODE REVIEW

**Comment by mikemorency** [comment]
recheck

File: changelogs/fragments/523-refactor-lookup-plugins.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+minor_changes:
+  - cluster_moid - Fix bug where lookup would return incosistent results for objects in nested paths (https://github.com/ansible-collections/vmware.vmware_rest/pull/523)
```
Comment (bardielle) [inline changelogs/fragments/523-refactor-lookup-plugins.yml:3]: please also mention the issue number.

**Review by bardielle — APPROVED** [review]

**Comment by bardielle** [comment]
@mariolenz can you review this PR please?

File: plugins/plugin_utils/lookup.py [hunk 2]
```diff
@@ -12,109 +12,90 @@
...
+                    filter_value
+                )
+            except KeyError:
+                corrected_filters[filter_key] = filter_value
+
+        return corrected_filters
+
+    async def fetch_object_with_filters(self, object_type, filters):
+        _url = self.build_url(object_type, filters)
+        res = await self.fetch(_url)
```
Comment (machacekondra) [inline plugins/plugin_utils/lookup.py:89]: why twice?

File: plugins/plugin_utils/lookup.py [hunk 3]
```diff
@@ -12,109 +12,90 @@
...
+        self.session = session
+
+    async def fetch(self, url):
+        async with self.session.get(url) as response:
+            result = await response.json()
+            return result
+
+    def build_url(self, object_type, filters):
+        corrected_filters_for_query = self.correct_filter_names(filters, object_type)
+        if object_type == "resource_pool":
```
Comment (machacekondra) [inline plugins/plugin_utils/lookup.py:57]: Maybe worth to make it generic?
Like `if '_' in object_type`?

**Review by machacekondra — APPROVED** [review]

File: plugins/plugin_utils/lookup.py (hunk 2 above)
Comment (mikemorency) [reply plugins/plugin_utils/lookup.py:89]: i think the old implementation of the plugin used both methods. Theres no reason for two different methods now though, ill combine them

File: plugins/plugin_utils/lookup.py (hunk 3 above)
Comment (mikemorency) [reply plugins/plugin_utils/lookup.py]: good point, i think theres no reason to even have the `if` statement too

**Pull request merged**